core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala

```diff
@@ -209,58 +191,45 @@ private[zio] final class FiberContext[E, A](
 
                   curIo = nextInstr(value)
 
-                case ZIO.Tags.Uninterruptible =>
-                  val io = curIo.asInstanceOf[ZIO.Uninterruptible[Any, E, Any]]
+                case ZIO.Tags.InterruptStatus =>
+                  val io = curIo.asInstanceOf[ZIO.InterruptStatus[Any, E, Any]]
+
+                  curIo = changeInterrupt(io.zio, io.flag)
 
-                  curIo = doNotInterrupt(io.zio)
+                case ZIO.Tags.CheckInterrupt =>
+                  val io = curIo.asInstanceOf[ZIO.CheckInterrupt[Any, E, Any]]
+
+                  curIo = io.f(interruptible)
 
                 case ZIO.Tags.Supervised =>
                   val io = curIo.asInstanceOf[ZIO.Supervised[Any, E, Any]]
 
+                  // TODO: Use bracket?
                   curIo = enterSupervision *> io.value.ensuring(exitSupervision)
 
                 case ZIO.Tags.Fail =>
-                  val io = curIo.asInstanceOf[ZIO.Fail[E]]
+                  val io = curIo.asInstanceOf[ZIO.Fail[E, Any]]
 
-                  val finalizer = unwindStack
+                  unwindStack()
 
                   if (stack.isEmpty) {
                     // Error not caught, stack is empty:
-                    if (finalizer eq null) {
-                      // No finalizer, so immediately produce the error.
-                      curIo = null
-
-                      done(Exit.halt(io.cause))
-                    } else {
-                      // We have finalizers to run. We'll resume executing with the
-                      // uncaught failure after we have executed all the finalizers:
-                      curIo = doNotInterrupt(finalizer).flatMap(
-                        cause => IO.halt(Option.option2Iterable(cause).foldLeft(io.cause)(_ ++ _))
-                      )
-                    }
+                    curIo = null
+
+                    done(Exit.halt(io.cause))
```

Comment:
Not sure if you want to fix #741 in this PR, but it is still relevant with this code.

Reply:
Right, I will try to fix that too. Did you already write a test for that that I can steal? Any other tests I can steal? 😄

Reply:
The #741 is a hard one as it is nondeterministic: it only happens when interruption occurs in the `ZIO.Fail(Cause.Fail)` case of the main loop before `unwindStack` finds the `ZIO.Fold`. Maybe run it 100 times or so. You seem to already have tests from #731. And #724 version of `interruptible` has semantic differences vs this PR.

Reply:
OK, I found a way to reproduce it and fixed it.